File: low_level/python/e7_voice_pub.py
#!/usr/bin/env python3
"""
Voice Command Publisher

Supports two modes:
1. file mode: publish a local audio file path once
2. streaming mode: capture audio from microphone in real-time and publish
"""

import sys
import subprocess
import time
import threading
import queue
import dds_middleware_python as dds


class AudioCaptureThread(threading.Thread):
    """Background thread for continuous low-latency audio capture (24kHz, mono, 16bit)."""

    def __init__(self, chunk_duration_ms=100):
        super().__init__(daemon=True)
        self.chunk_duration_ms = chunk_duration_ms
        self.audio_queue = queue.Queue(maxsize=2)  # small buffer to avoid lag
        self.running = True

    def run(self):
        process = subprocess.Popen(
            ["arecord", "-q", "-t", "raw", "-f", "S16_LE", "-c", "1", "-r", "24000"],
            stdout=subprocess.PIPE,
            bufsize=0,  # unbuffered for lowest latency
        )
        # 24kHz * 16bit mono = 2 bytes/sample
        bytes_per_chunk = int(24000 * self.chunk_duration_ms / 1000 * 2)
        try:
            while self.running:
                chunk = process.stdout.read(bytes_per_chunk)
                if not chunk:
                    break
                # Drop the oldest chunk if the queue is full to stay real-time.
                if self.audio_queue.full():
                    self.audio_queue.get_nowait()
                self.audio_queue.put_nowait(bytearray(chunk))
        except Exception as e:
            print(f"Audio capture thread error: {e}")
        finally:
            process.terminate()

    def get_audio(self):
        """Return the next available audio chunk, or None if the queue is empty."""
        try:
            return self.audio_queue.get_nowait()
        except queue.Empty:
            return None

    def stop(self):
        self.running = False


def make_header():
    """Build a std_msgs Header stamped with the current time.

    The refactored VoiceCmd (dds-middleware >= 0.23.x) carries a Header,
    matching the canonical voice publisher examples.
    """
    header = dds.Header()
    stamp = dds.Time()
    now = time.time()
    stamp.sec(int(now))
    stamp.nanosec(int((now - int(now)) * 1e9))
    header.stamp(stamp)
    header.frame_id("voice_cmd")
    return header


def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "file"

    middleware = dds.PyDDSMiddleware(0)
    qos_config = {
        "reliability": "reliable",
        "history_kind": "keep_last",
        "history_depth": 5,
        "durability": "volatile",
    }
    middleware.createVoiceCmdWriter("rt/voice/cmd", qos_config)

    print(f"Mode: {mode}, QoS: RELIABLE, KEEP_LAST(5), VOLATILE")

    if mode == "file":
        file_path = "/root/test2.flac"

        voice_cmd = dds.VoiceCmd()
        voice_cmd.header(make_header())
        voice_cmd.priority(dds.VoicePriority.kNormal)  # kNormal: ordinary audio file
        voice_cmd.task_id("e7_voice_pub")
        voice_cmd.type("file")
        voice_cmd.path(file_path)
        voice_cmd.data([])
        voice_cmd.flag(False)  # stream-end flag, unused in file mode

        time.sleep(1)  # wait for DDS discovery
        middleware.publishVoiceCmd(voice_cmd)
        print(f"Published VoiceCmd (file): {voice_cmd.path()}")

    elif mode == "streaming":
        print("Streaming mode: capture and publish from microphone (low-latency)")

        capture_thread = AudioCaptureThread(chunk_duration_ms=100)
        capture_thread.start()

        try:
            while True:
                audio = capture_thread.get_audio()
                if audio is None:
                    time.sleep(0.01)
                    continue

                voice_cmd = dds.VoiceCmd()
                voice_cmd.header(make_header())
                voice_cmd.priority(dds.VoicePriority.kNormal)  # kNormal: ordinary audio stream
                voice_cmd.task_id("e7_voice_pub")
                voice_cmd.type("streaming")
                voice_cmd.path("")
                voice_cmd.data(list(audio))
                voice_cmd.flag(False)  # False: stream not finished yet

                middleware.publishVoiceCmd(voice_cmd)
                print(f"Published VoiceCmd (streaming): {len(voice_cmd.data())} bytes")

        except KeyboardInterrupt:
            print("Stopping streaming...")
        finally:
            capture_thread.stop()

    else:
        print("Unknown mode, use 'file' or 'streaming'")


if __name__ == "__main__":
    main()

File: low_level/cpp/e7_voice_pub.cc
#include "dds_middleware.hpp"
#include "voice_cmd.hpp"
#include <atomic>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <iostream>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

using namespace dds_middleware;
using namespace dobotmh4::msg::dds_;

// Populate the std_msgs Header with the current wall-clock timestamp.
// The refactored VoiceCmd_ IDL (dds-middleware >= 0.23.x) carries a Header,
// matching the canonical voice_cmd_publisher example.
static void fill_header(VoiceCmd_& voice_cmd)
{
    auto now = std::chrono::system_clock::now().time_since_epoch();
    auto sec = std::chrono::duration_cast<std::chrono::seconds>(now).count();
    auto nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(now).count() % 1000000000;
    voice_cmd.header().stamp_().sec_(static_cast<int32_t>(sec));
    voice_cmd.header().stamp_().nanosec_(static_cast<uint32_t>(nsec));
    voice_cmd.header().frame_id_("voice_cmd");
}

// Background thread that continuously captures PCM audio from the microphone
// (24kHz, mono, 16bit) and keeps only the most recent chunk for low latency.
class AudioCaptureThread
{
public:
    explicit AudioCaptureThread(int chunk_duration_ms = 100)
        : chunk_duration_ms_(chunk_duration_ms)
        , running_(true)
    {
    }

    void run()
    {
        FILE* pipe = popen("arecord -q -t raw -f S16_LE -c1 -r24000", "r");
        if (!pipe) {
            std::cerr << "arecord not found or failed to start" << std::endl;
            return;
        }

        // 24kHz * 16bit mono = 2 bytes/sample, e.g. 100ms -> 4800 bytes
        int bytes_per_chunk = (24000 * chunk_duration_ms_ / 1000) * 2;
        std::vector<uint8_t> buffer(bytes_per_chunk);

        while (running_) {
            size_t read_bytes = fread(buffer.data(), 1, bytes_per_chunk, pipe);
            if (read_bytes > 0) {
                std::lock_guard<std::mutex> lock(mutex_);
                latest_.assign(buffer.begin(), buffer.begin() + read_bytes);
                has_data_ = true;
            } else if (feof(pipe)) {
                break;
            }
        }

        pclose(pipe);
    }

    bool get_audio(std::vector<uint8_t>& data)
    {
        std::lock_guard<std::mutex> lock(mutex_);
        if (!has_data_) {
            return false;
        }
        data = std::move(latest_);
        has_data_ = false;
        return true;
    }

    void stop() { running_ = false; }

private:
    int chunk_duration_ms_;
    std::atomic<bool> running_;
    std::mutex mutex_;
    std::vector<uint8_t> latest_;
    bool has_data_ = false;
};

int main(int argc, char** argv)
{
    std::string mode = (argc > 1) ? argv[1] : "file"; // "file" or "streaming"

    auto middleware = std::make_shared<DDSMiddleware>(0);

    QoSProfile qos;
    qos.reliability = ReliabilityPolicy::RELIABLE;
    qos.history = HistoryPolicy::KEEP_LAST;
    qos.history_depth = 5;
    qos.durability = DurabilityPolicy::VOLATILE;

    auto publisher = middleware->create_publisher<VoiceCmd_>("rt/voice/cmd", qos);

    std::cout << "Mode: " << mode << ", QoS: RELIABLE, KEEP_LAST(5), VOLATILE" << std::endl;

    if (mode == "file") {
        std::string file_path = "/root/test2.flac";

        VoiceCmd_ voice_cmd;
        fill_header(voice_cmd);
        voice_cmd.priority(VoicePriority_::kNormal); // kNormal: ordinary audio file
        voice_cmd.task_id("e7_voice_pub");
        voice_cmd.type("file");
        voice_cmd.path(file_path);
        voice_cmd.data().clear();
        voice_cmd.flag(false); // stream-end flag, unused in file mode

        std::this_thread::sleep_for(std::chrono::seconds(1)); // wait for DDS discovery
        publisher->publish(voice_cmd);
        std::cout << "Published VoiceCmd (file): " << voice_cmd.path() << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1)); // let the sample flush before exit
    } else if (mode == "streaming") {
        std::cout << "Streaming mode: capture and publish from microphone (low-latency)" << std::endl;

        AudioCaptureThread capture_thread(100); // 100ms chunks
        std::thread audio_thread(&AudioCaptureThread::run, &capture_thread);

        std::vector<uint8_t> audio;
        while (true) {
            if (!capture_thread.get_audio(audio)) {
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
                continue;
            }

            VoiceCmd_ voice_cmd;
            fill_header(voice_cmd);
            voice_cmd.priority(VoicePriority_::kNormal); // kNormal: ordinary audio stream
            voice_cmd.task_id("e7_voice_pub");
            voice_cmd.type("streaming");
            voice_cmd.path("");
            voice_cmd.data(audio);
            voice_cmd.flag(false); // false: stream not finished yet

            publisher->publish(voice_cmd);
            std::cout << "Published VoiceCmd (streaming): " << voice_cmd.data().size() << " bytes" << std::endl;
        }

        capture_thread.stop();
        audio_thread.join();
    } else {
        std::cout << "Unknown mode, use 'file' or 'streaming'" << std::endl;
    }

    return 0;
}
